### ISSUE TITLE
 Error when checking input: expected Length to have 1 dimensions, but got array with shape (10, 1)

### ISSUE THREAD

**Issue by zhangyi02** [keras-team/keras#12058]
In my self_build model,  input layer was defined as below:
`len_input = Input(batch_shape=(None,), dtype=tf.int64, name='Length')`
Then I try to run `model.train_on_batch([mix, mtx, leng], src)` (mix, mtx, and leng are 3 input, src is the ground truth, and mix, mtx, src has no concern about this problem)
But got this Error:

> Error when checking input: expected Length to have 1 dimensions, but got array with shape (10, 1)

And when I print `leng.shape`, got `(10,)` with 10 as my batch_size.

So, here is the problem, I put a (?,) dimension array in an array list as model input, but the model parse it as (?,1) dimension and got an Error.
Or to say, is there other way to give a model multi inputs?

**Comment by ParikhKadam** [comment]
@zhangyi02 

I will need to see the full code of the model in order to solve this.. Or maybe some more amount of code than this.

There are two ways to pass multiple inputs to the model:
1. As a list
2. As a dictionary

**Comment by zhangyi02** [comment]
@ParikhKadam @msymp 
Simple code to build model:
```python
len_input = Input(batch_shape=(None,), name='Length')
leng = Lambda(lambda x: x*2, trainable=False)(len_input)
model = Model(inputs=len_input, outputs=leng)
model.compile(loss='mse', optimizer='adam')
# feed some data
a = np.array([1])
b = np.array([2])
model.train_on_batch(a, b)
```
Then got error as :
 
> Error when checking input: expected Length to have 1 dimensions, but got array with shape (1, 1)

While it's so clear a's shape is `(1,)` not `(1, 1)`.
However, when I re-produce this error, I found that I misunderstood what's going wrong. It's nothing about the input list, as just one input used in the example, maybe it's all about `batch_shape=(None, )`.

**Comment by ParikhKadam** [comment]
@zhangyi02 
The problem was hard to trace but I think it's solved...

The Input layer in keras works differently than expected. It will always return a tensor of rank 2 or more. By default, the 1st dimension is `batch_size `and then other dimensions are `features`. Here, you tried to make a model with features of empty shape i.e. 0 rank. In short, you tried to pass a scalar value as input to the neural network (ignoring the batches). This is not possible as neural networks are generalized to work on tensors.

Hence, the `Input` layer in keras does the task of expanding dimensions around your scalar valued features/input. That means, if your 1st sample contains a **single feature x1=2**, the input layer will not generate the tensor for 1st sample as **2** (because it's a **scalar**). Rather, it will generate a **tensor** for 1st sample as **[2]**. This will definitely have **shape=(1, )**.

Now, combine batches of samples and it will add `batch_size` to the shape mentioned above which will give you **(None, 1)**.

In short, the output of Input layer will always have shape **_(batch_size, no. of features) + shape_of_feature_**. Here, number of features is **1** and shape of that single feature is **()**. Hence, you will get (None, 1).

-- The documentation didn't contain this but I am sure the Input layer works this way --

To make your code work, replace

`len_input = Input(batch_shape=(None,), name='Length')`

with

`len_input = Input(batch_shape=(None, 1), name='Length')`

You can ask the developers regarding such a behaviour of Input layer and also ask if you want to update the documentation.

Thanks...

**Comment by zhangyi02** [comment]
@ParikhKadam 
I guess so, thanks a lot for your explanation.

**Comment by ParikhKadam** [comment]
@zhangyi02 
Welcome...

**Comment by msymp** [comment]
This issue is closed. Thanks @ParikhKadam .

**Comment by ParikhKadam** [comment]
Welcome.. @msymp
